include/kernel.h

```diff
@@ -4401,7 +4401,7 @@ struct k_msgq_attrs {
 #define K_MSGQ_DEFINE(q_name, q_msg_size, q_max_msgs, q_align)		\
 	static char __noinit __aligned(q_align)				\
 		_k_fifo_buf_##q_name[(q_max_msgs) * (q_msg_size)];	\
-	STRUCT_SECTION_ITERABLE(k_msgq, q_name) =			\
+	static STRUCT_SECTION_ITERABLE(k_msgq, q_name) =		\
```

Comment:
@carlescufi Yep, that's right - `K_PIPE_DEFINE` is also affected. I have just walked through `kernel.h` and all occurrences of `STRUCT_SECTION_ITERABLE` and we should actually introduce `#define <...>_DEFINE_STATIC` suffixes to all helper macros, where there is no solution to just prepend `static` in front of those (i.e. they define some auxiliary buffer before the actual object).

I have made a similar PR before (already merged #39683):

* `K_MEM_SLAB_DEFINE` - Statically define and initialize a memory slab in a public (non-static) scope.
* `K_MEM_SLAB_DEFINE_STATIC` - Statically define and initialize a memory slab in a private (static) scope.

So, this original commit is flawed in a way it prevents to `extern` such definition elsewhere. Although not the best way, still the API could keep this option available.

Based on this logic, we can discuss the availability of:

* `K_MSGQ_DEFINE`
* `K_MSGQ_DEFINE_STATIC`
* `K_PIPE_DEFINE`
* `K_PIPE_DEFINE_STATIC`
* etc.

But for `K_MUTEX_DEFINE`, it is not necessary, as `static K_MUTEX_DEFINE` does the job. The other perspective: For the sake of API clarity, all `_DEFINE` macros should have their `_DEFINE_STATIC` counterparts.

What do you think? Thanks.

Reply:
Thanks @hubpav, I will let @andyross, @dcpleung and @peter-mitsis weigh in here.

Reply:
As you correctly point out, the K_MSGQ_DEFINE() macro defines two objects--a buffer and the k_msgq instance. It is logical for the buffer portion to be marked as 'static' (or private to the file). However, the k_msgq instance may be referenced from another file, so it is not marked as such.

I think that you are also correct in pointing out the need for a static (private to the file) instance of the k_msgq object. Since the instance created by K_MSGQ_DEFINE() is meant to in theory be available to other modules, I think your idea of creating a new macro K_MSGQ_DEFINE_STATIC() is the best route forward--particularly since there is precendent with K_MEM_SLAB_DEFINE_STATIC().